docs/config.md
# config

## What?

the configuration is split into 2 parts:
- wrauth db, which is a `db.yaml` file.
- wrauth configuration, which is a `config.yaml` file.

both files **must be in the same directory**. they are parsed on program start and on change and their position can be specified using command line arguments (check `wrauth --help`).  

all IP addresses **must** be in CIDR notation (have a subnet mask at the end, use `/32` as an equivalent to no subnet).

options with `EITHER:` and `OR:` are not mutually exclusive, i.e. `rules` entries can have both `ips` and `pubkeys`, but must have atleast one of them (thats why the prefix).

wrauth doesn't watch the Authelia user database, but it parses that on file change. to reload it, just write something (like a comment) to one of the watched `yaml` files and it will be reloaded.

IP rules are matched sequentially, but access rules aren't. it's in the order:
- check for direct matches
- check for bypasses
- check all domains with globbing, if any match

### wrauth db

```yaml
# REQUIRED: IP to user matching rules
rules:
    # EITHER: the addresses to match
  - ips:
    - '10.0.0.0/30'
    - '10.0.0.10/32'
    # REQUIRED: the user to designate
    user: 'alice'
    # OR: the public keys (from WireGuard) to match
  - pubkeys: [ 'MJ6JoquFLTf419V5dzkcV1z8TY8SIuPyaSH/1SBBP1o=' ]
    user: 'bob'

# REQUIRED: access control rules
access:
  # REQUIRED: the domains to match
  - domains: [ 'private.example.com' ]
    # OPTIONAL: regex to match for path
    # NOTE: https://github.com/google/re2/wiki/Syntax
    resource: '/(cpanel|database).*'
    # EITHER: the users to allow
    users: [ 'admin' ]
    # OR: the groups to allow
    groups:
      - [ 'trusted' ]
  # NOTE: https://pkg.go.dev/path/filepath#Match
  # NOTE: you CANNOT bypass a glob, this is by design
  - domains: [ '*.example.com' ]
    users: [ 'superadmin' ]
  - domains: [ 'public.example.com' ]
    # NOTE: allow all users (this is how you bypass)
    users: [ '*' ]
  - domains: 
    - 'test.example.com'
    - 'devdb.example.com'
    users: 
      - 'databaseguy'
      - 'admins'
    groups:
      # NOTE: matches users who are in 'maindbs' or in both 'sys' and 'devs'
      - - 'maindbs'
      - - 'devs' 
        - 'sys'
    # OPTIONAL: any site specific headers to add
    headers:
      # OPTIONAL: header name: header value
      X-AuthDB-Roles: "devdb"

# REQUIRED: site admin rules
admins:
  users: [ 'admin' ]
  groups:
    - - 'admins'
    - - 'sys'
      - 'trusted'
```

### wrauth configuration

```yaml
# OPTIONAL: the port to listen on.
# NOTE: wrauth doesn't support ipv6 (and doesn't plan to)
# NOTE: wrauth only listens on 127.0.0.1. this is by design, it is not meant to be used outside of a reverse proxy.
# DEFAULT: 9092
# NORELOAD
address: '9093'
# REQUIRED: the external address, https is assumed
external: 'wrauth.example.com'
# OPTIONAL: enable or disable caching
# NOTE: this will reduce performance SIGNIFICANTLY if disabled, since proxying Authelia directly reduces performance by ~40%
# DEFAULT: true
# NORELOAD
caching: true
# OPTIONAL: the log level 
# DEFAULT: info
# NORELOAD
level: 'debug'
# OPTIONAL: the theme (currently only gruvbox-dark)
# DEFAULT: gruvbox-dark
theme: 'gruvbox-dark'

# REQUIRED: Authelia configuration
authelia:
  # REQUIRED: Authelia's listening address
  address: '127.0.0.1:9091'
  # REQUIRED: Authelia's user database
  db: '/opt/authelia/users.yaml'
  # OPTIONAL: no. connections to keep open with Authelia
  # DEFAULT: 64
  # NORELOAD
  connections: 32
  # OPTIONAL: how often to clear cache in seconds
  # DEFAULT: 300
  # NORELOAD
  cache: 600
  # OPTIONAL: how often to ping Authelia in seconds (must be below 30 to keep connections alive)
  # DEFAULT: 25
  # NORELOAD
  ping: 28

# REQUIRED: WireGuard interfaces
interfaces:
  # REQUIRED: interface name
  - name: 'wg0'
    # REQUIRED: listening address
    addr: '10.0.0.1/32'
    # OPTIONAL: the configuration file
    # DEFAULT: /etc/wireguard/<name>.conf
    conf: '/etc/wireguard/wg.conf'
  - name: 'wg1'
    addr: '172.16.0.1'
    # OPTIONAL: time from the last handshake to consider a connection closed
    # DEFAULT: 150
    shake: 300
```

### Authelia configuration

this is just the recommended Authelia configuration demonstration how it should be used with wrauth. notice how **there are no conflicting network based rules**, this is important, so that there are no conflicts in authentication, e.g. wrauth disallows or doesn't match `10.0.0.5` (intended) but Authelia has bypassed that entire network (accidental), so wrauth requests Authelia and it immediately responds with 200 OK, causing wrauth to reply with that.

```yaml
  rules:
    # public, allow everybody
    - domain:
      - 'example.com'
      - 'www.example.com'
      - 'public.example.com'
      policy: 'bypass'
    # semi-private, allow everybody who's either on the VPN or authenticated
    - domain: [ 'semi.example.com' ]
      policy: 'bypass'
      networks: 'vpn'
    - domain: [ 'semi.example.com' ]
      policy: 'two_factor'
    # private, allow only specific users
    - domain: '*.example.com'
      policy: 'two_factor'
      subject: 'group:admins' 
    - domain: [ 'db.example.com', 'git.example.com' ]
      policy: 'two_factor'
      subject: 'group:devs'
```


conf.go
package main

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/fsnotify/fsnotify"
	"go.uber.org/zap"
	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
)

func ParseFiles() error {
	if err := ParseYaml(&Conf, Args.Config); err != nil {
		return fmt.Errorf("configuration: %w", err)
	}
	if err := ParseYaml(&Db, Args.DB); err != nil {
		return fmt.Errorf("database: %w", err)
	}
	if err := ParseYaml(&Db, Conf.Authelia.Db); err != nil {
		return fmt.Errorf("Authelia user database: %w", err)
	}

	return nil
}

func CheckConf() error {
	if Conf.External == "" {
		return fmt.Errorf("external address not configured")
	}
	if Conf.Authelia.Address == "" {
		return fmt.Errorf("Authelia address not configured")
	}
	if Conf.Authelia.Db == "" {
		return fmt.Errorf("Authelia user database path not configured")
	}
	if Conf.Authelia.Ping >= 30 {
		return fmt.Errorf("configured ping interval is too large")
	}
	if len(Conf.Interfaces) == 0 {
		return fmt.Errorf("interfaces not configured")
	}

	for i, inf := range Conf.Interfaces {
		if inf.Name == "" {
			return fmt.Errorf("interface name not configured")
		}
		if inf.Addr.Mask == 0 {
			return fmt.Errorf("address not configured for interface %v", inf.Name)
		}
		if inf.Conf == "" {
			Conf.Interfaces[i].Conf = "/etc/wireguard/" + inf.Name + ".conf"
		}
		if inf.Shake == 0 {
			Conf.Interfaces[i].Shake = 150
		}
	}

	return nil
}

func CheckDB() error {
	for _, r := range Db.Rules {
		if r.User == "" {
			return fmt.Errorf("rules: user not configured")
		}
		if len(r.Ips) == 0 && len(r.Pubkeys) == 0 {
			return fmt.Errorf("rules: IPs or public keys not configured for %v", r.User)
		}
	}
	for _, a := range Db.Access {
		if len(a.Domains) == 0 {
			return fmt.Errorf("access: domains not configured")
		}
		if len(a.Users) == 0 && len(a.Groups) == 0 {
			return fmt.Errorf("access: neither users nor groups configured for %v", a.Domains)
		}
	}
	if len(Db.Admins.Users) == 0 && len(Db.Admins.Groups) == 0 {
		return fmt.Errorf("admins not configured")
	}

	return nil
}

func WatchFS(w *fsnotify.Watcher) {
	for {
		select {
		case err, ok := <-w.Errors:
			if !ok {
				return
			}
			Log.Fatalln("file watch:", err)
		case ev, ok := <-w.Events:
			if !ok {
				return
			}
			if ev.Op == fsnotify.Write && (strings.Contains(ev.Name, Args.Config) || strings.Contains(ev.Name, Args.DB)) {
				Log.Infoln("file changed:", ev.Name)
				Clear()
				if err := LoadFiles(); err != nil {
					Log.Fatalln("loading files:", err)
				}
				if err := LoadData(); err != nil {
					Log.Fatalln("processing data:", err)
				}
			}
		}
	}
}

func Clear() {
	Conf = Config{
		Address: "127.0.0.1:9092",
		Level:   zap.NewAtomicLevel(),
		Caching: true,
		Theme:   "gruvbox-dark",
		Authelia: Authelia{
			Connections: 64,
			Cache:       300,
			Ping:        25,
		},
	}
	Db = DB{}
	Matches = nil
	Cache = make(map[string]map[string][]byte)
	Regexps = make(map[string]map[string]*regexp.Regexp)
}

func LoadFiles() error {
	if err := ParseFiles(); err != nil {
		return fmt.Errorf("parsing: %w", err)
	}
	Log.Debugln("checking configuration")
	if err := CheckConf(); err != nil {
		return fmt.Errorf("configuration: %w", err)
	}
	if !Conf.Caching {
		Log.Warnln("caching is disabled for all Authelia requests")
	}
	Log.Debugln("checking database")
	if err := CheckDB(); err != nil {
		return fmt.Errorf("database: %w", err)
	}
	return nil
}

func LoadData() error {
	Log.Debugln("adding cache")
	if err := AddCache(); err != nil {
		return fmt.Errorf("caching: %w", err)
	}

	var infs []*wgtypes.Device
	for _, inf := range Conf.Interfaces {
		dev, err := WGClient.Device(inf.Name)
		if err != nil {
			return fmt.Errorf("WireGuard device %v: %w", inf.Name, err)
		}
		if dev.Type != wgtypes.LinuxKernel {
			Log.Warnf("wrauth is using userspace WireGuard device %v", inf.Name)
		}

		Log.Debugln("adding interface:", inf.Name)
		infs = append(infs, dev)
	}

	Log.Debugln("adding IP matches")
	// needs WireGuard setup
	if err := AddMatches(infs); err != nil {
		return fmt.Errorf("matching: %w", err)
	}
	return nil
}
